src/arrays/quarray.jl

```diff
@@ -0,0 +1,148 @@
+import Base: size,
+    length,
+    getindex,
+    similar,
+    in,
+    ctranspose,
+    transpose,
+    summary,
+    zeros,
+    eye,
+    #TODO: Implement the below operations
+    *,.*,
+    /,./,
+    +,.+,
+    -,.-,
+    kron
+
+abstract AbstractQuArray{B<:(AbstractBasis...), T, N} <: AbstractArray{T,N}
+
+###########
+# QuArray #                 
+###########
+    # B<:(AbstractBasis...) allows us to have a basis for each dimension, 
+    # and gives us a less ambiguous way to determine if a QuArray will act
+    # like a vector, matrix, or tensor compared to just looking at the 
+    # dimension parameter N. 
+    type QuArray{B<:(AbstractBasis...), T, N, A} <: AbstractQuArray{B, T, N}
+        coeffs::A
+        bases::B
+        function QuArray(coeffs::AbstractArray{T, N}, bases::B) 
+            if checkbases(coeffs, bases) 
+                new(coeffs, bases)
+            else 
+                error("Coefficient array does not conform to input bases")
+            end
+        end
+    end
+
```

Comment:
So now that we know about `NTuple`s, how do folks feel about changing `QuArray` to this:

```
abstract AbstractQuArray{B<:AbstractBasis, T, N} <: AbstractArray{T,N}

type QuArray{B<:AbstractBasis, T, N, A} <: AbstractQuArray{B, T, N}
    coeffs::A
    bases::NTuple{N, B}
    function QuArray(coeffs::AbstractArray{T}, bases::NTuple{N, B}) 
        if checkbases(coeffs, bases) 
            new(coeffs, bases)
        else 
            error("Coefficient array does not conform to input bases")
        end
    end
end
```

I believe the above to be much, much cleaner. There are two significant changes with this approach:

1. There can only be a single basis type for the entire Array. I don't really see any detriment to 
this - when would you ever want different bases types along different dimensions? If anything, the 
restriction makes the whole object easier to reason about.

2. We define the number of dimensions of a `QuArray` to be equivalent to how many bases its storing. 
This way, `QuArray{..., 1}` really does act like a vector, `QuArray{..., 2}` really does act like an
operator, and `QuArray{..., 3+}` really does act like a tensor. This approach allows us to sidestep 
some of the issues with the conflation of vectors and matrices. I was already heading down this road
with the `QuVector` and `QuMatrix` type aliases, but this is much more clearly laid out.